csrc/multidevice/utils.cpp

```diff
@@ -141,20 +141,53 @@ int64_t numDeviceDims(TensorView* tv) {
       [](IterDomain* id) { return id->isDeviceDim(); });
 }
 
-bool isResharding(Expr* expr) {
-  std::unordered_set<TensorView*> tvs;
-  for (auto tv : ir_utils::filterByType<TensorView>(expr->inputs())) {
-    tvs.insert(tv);
+bool haveDifferentShardings(TensorView* producer, TensorView* consumer) {
+  // exit early in the unsharded case for performance
+  if (!producer->hasDeviceMesh() && !consumer->hasDeviceMesh()) {
+    return false;
   }
-  for (auto tv : ir_utils::filterByType<TensorView>(expr->outputs())) {
-    tvs.insert(tv);
+  // If device mesh are different, the Expr is resharding
+  if (!(producer->getDeviceMesh() == consumer->getDeviceMesh())) {
+    return true;
   }
-  if (tvs.empty()) {
-    return false;
+  // Create a map between producer's and consumer's IterDomains. We iterate
+  // over producer's iterdomain and compare sharding type with consumer's
+  // iterdomain
+  const auto p2c_map =
+      PairwiseRootDomainMap(producer, consumer).mapProducerToConsumer();
+  for (auto p_id :
+       TensorDomain::noReductions(producer->getMaybeRFactorDomain())) {
+    auto p2c_map_it = p2c_map.find(p_id);
+    NVF_ERROR(
+        p2c_map_it != p2c_map.end(),
+        "the producer ",
+        producer,
+        " has a dimension ",
+        p_id,
+        " that is not mapped to its consumer ",
+        consumer);
+    auto c_id = p2c_map_it->second;
+    if (p_id->getParallelType() != c_id->getParallelType() &&
+        (p_id->isDeviceDim() || c_id->isDeviceDim())) {
```

Comment:
```suggestion
    if (p_id->isDeviceDim != c_id->isDeviceDim) {
```

I think this is equivalent and simpler.

Reply:
I'd rather keep it the way it was so we won't have to change it when we soon introduce the new `DIDy` parallel type